onnxruntime/core/providers/cpu/math/top_k.cc

```diff
@@ -45,92 +47,175 @@ static int64_t SizeFromDim(size_t k, const vector<int64_t>& dims) {
+                                   const unsigned k, const unsigned axis_parsed) {
+  // Cache some values that will be used in the implementation below
+  const int64_t rows = SizeToDim(axis_parsed, input_shape);
+  const int64_t cols = input->Shape().Size() / rows;
   auto input_map = ConstEigenMatrixMapRowMajor<float>(
-      static_cast<const float*>(X->template Data<float>()),
+      static_cast<const float*>(input->template Data<float>()),
       rows,
       cols);
 
-  // Resize output tensors to be the same shape as the input except
-  // for the specified dimension ((i.e.) axis_parsed), which will be of size k. E.x. for an input tensor
-  // of shape [3, 4, 5] and k=2 with axis_parsed=1, both of these will be shape [3, 2, 5]
-  vector<int64_t> output_linear_shape = in_dims;
-  output_linear_shape[axis_parsed] = k;
-  auto* Values = p_op_kernel_context->Output(0, output_linear_shape);
-  auto* Indices = p_op_kernel_context->Output(1, output_linear_shape);
-
   // Use Eigen maps to allow indexing into the 2d tensors like Values_map(i,j)
-  const int64_t reduced_cols = SizeFromDim(axis_parsed, output_linear_shape);
-  auto Values_map = EigenMatrixMapRowMajor<float>(
-      Values->template MutableData<float>(), rows, reduced_cols);
-  auto Indices_map = EigenMatrixMapRowMajor<int64_t>(
-      Indices->template MutableData<int64_t>(), rows, reduced_cols);
+  const int64_t reduced_cols = SizeFromDim(axis_parsed, output_shape);
+  auto values_map = EigenMatrixMapRowMajor<float>(
+      values->template MutableData<float>(), rows, reduced_cols);
+  auto indices_map = EigenMatrixMapRowMajor<int64_t>(
+      indices->template MutableData<int64_t>(), rows, reduced_cols);
 
   // This is basically the number of elements within each of the "k" rows
   const int64_t block_slice = reduced_cols / k;
-  // Sort preserving Indices
+  const int64_t num_blocks = input_shape[axis_parsed];
+  // Sort preserving indices
   for (int64_t i = 0; i < rows; ++i) {
     for (int64_t j = 0; j < block_slice; ++j) {
-      // Build a min-heap, the heap element is pair of (value, idx)
-      // the top of the heap is the smallest value
-      priority_queue<
-          pair<float, int64_t>,
-          vector<pair<float, int64_t>>,
-          ValueCmp<float>>
-          min_heap;
-      // Maintain the size of heap to be less or equal to k_, so the
-      // heap will hold the k largest Values
-      for (int64_t l = 0; l < in_dims[axis_parsed]; ++l) {
-        const auto value = input_map(i, l * block_slice + j);
-        if (min_heap.size() < k || value > min_heap.top().first) {
-          min_heap.push({value, l});
+      // If the request is to sort the TopK values, we will use a Heap to
+      // hold the top K values in sorted fashion
+      if (sorted) {  // the optimizer will clean-up the redundant condition based on the template parameter 'sorted'
+        // Build a min-heap/max-heap, the heap element is pair of (value, idx)
```

Comment:
Is there perf benefit from separating sorted vs unsorted during selection, or would simply selecting the top K first and sorting that later if needed be simpler?

Reply:
Hmm, I did think about it but didn't lean one way or the other. 

If we walk this through - 

1) I think selecting top k elements without sorting should be AVERAGE O(n) and then sorting the k elements - k log(k) 

2) If we directly pick the sorted top k elements it would be n log(k) - heap of size 'k' and pass over the N elements while performing heap operations.

I think (1) MIGHT be better on average but worst-case perf might be poorer. 

Ultimately, I kept the code the same for opset-9, opset-10 (and opset-11 sorted == 1).

Reply:
What is your suggestion ? It should be trivial code tweak to pick first and sort (if needed) later.

Reply:
The typical use-case would be that `k` is much smaller than `n` (e.g. select top 3 categories out of 1000). Since the [nth_element()](https://en.cppreference.com/w/cpp/algorithm/nth_element) function has linear O(n) performance,  it would be noticeably faster to select top `k` and sort the k elements if needed.

However, since this is a good operator to performing sorting, someone might construct a model where k ==n,  simply to sort the results. In this scenario, it's always faster to skip the nth_element() function selection and  use heap sort for all elements.

One option would be to test if `n + k log k < n log n`. If yes, select top-k elements first and then sort them. If no, skip the nth_element() selection, and sort the entire array.

Otherwise a default of select first and sort later (if needed) seems like a safe choice.

Reply:
Thanks Jignesh. That looks like a good approach. Although, shouldn't it be:

`n + k log (k) < n log (k)` (as opposed to n log (n)). We don't sort the entire input array (n log(n)). We keep the heap size to be `k` (cost of a heap insert/delete is thus log (k)) and pass over the `n` input elements making it n log(k).

Reply:
Yes, the right hand side should be  `n log (k)` -- since the heap only stores k elements.

Reply:
Added 2 different approaches while picking sorted results -

1) If `n + k * ln k < n * ln k`: Select first and then sort the chosen `k` elements next - O(n + k * ln k)

2) Else: Use the sorted selection approach - O(n * ln k)

Added 2 tests that picks sorted values in the 2 different approaches as listed above